ggshield/sca/output/text_handler.py

```diff
@@ -1,20 +1,150 @@
+from collections import Counter
 from io import StringIO
-from typing import ClassVar, Optional
+from typing import ClassVar, Dict, Generator, List, NamedTuple, Optional
 
 from ggshield.core.constants import IncidentStatus
-from ggshield.core.text_utils import STYLE, file_info, format_text
+from ggshield.core.text_utils import STYLE, file_info, format_text, pluralize
 from ggshield.sca.collection import (
     SCAScanAllVulnerabilityCollection,
     SCAScanDiffVulnerabilityCollection,
 )
 from ggshield.sca.output.handler import SCAOutputHandler
 from ggshield.sca.sca_scan_models import (
     SCALocationVulnerability,
-    SCAVulnerability,
-    SCAVulnerablePackageVersion,
+    SCAScanAllOutput,
+    SCAScanDiffOutput,
 )
 
 
+class OutputIncidentData(NamedTuple):
+    package_full_name: str
+    version: str
+    severity: str
+    summary: str
+    fixed_version: Optional[str]
+    identifier: str
+    cve_ids: List[str]
+
+    def sort_key(self) -> int:
+        if self.severity == "critical":
+            return 0
+        if self.severity == "high":
+            return 10
+        if self.severity == "medium":
+            return 20
+        if self.severity == "low":
+            return 30
+        return 100
+
+
+class OutputLocationData(NamedTuple):
+    """
+    Aggregate SCA incidents data to be printed in one class
+    The counters allow to sort according to the number of incidents with each severity
+    """
+
+    location: str
+    added: List[OutputIncidentData]
+    removed: List[OutputIncidentData]
+    added_counter: Dict[str, int]
+    removed_counter: Dict[str, int]
+
+    def sort_key(self) -> List[int]:
+        return list(self.added_counter.values()) + list(self.removed_counter.values())
+
+
+def populate_incidents_list(
+    result_list: List[SCALocationVulnerability],
+    locations_with_incidents: Dict[str, OutputLocationData],
+    incident_list_key: str,
+    counter_key: str,
+) -> None:
+    for vuln_dep_file in result_list:
+        if vuln_dep_file.location not in locations_with_incidents:
+            locations_with_incidents[vuln_dep_file.location] = OutputLocationData(
+                location=vuln_dep_file.location,
+                added=[],
+                removed=[],
+                added_counter={"critical": 0, "high": 0, "medium": 0, "low": 0},
+                removed_counter={"critical": 0, "high": 0, "medium": 0, "low": 0},
+            )
+
+        # Update counter of vulns to sort locations by number of incidents
+        for pkg_version_with_vuln in vuln_dep_file.package_vulns:
+            for vuln in pkg_version_with_vuln.vulns:
+                incidents_list = locations_with_incidents[
+                    vuln_dep_file.location
+                ].__getattribute__(incident_list_key)
+                incidents_list.append(
+                    OutputIncidentData(
```

Comment:
Wouldn't it be easier if you use only dictionaries instead of `OutputLocationData` and `OutputIncidentData` and then at the end serialize those into the models?
Then for example if you specify a default_value for `removed` and `removed_counter` you would not need to put them in the dictionaries.
WDYT?

Reply:
I agree with the default values, I fixed it. About the dict, I don't think it would change anything. Where do you think it would get easier ?